### DISCUSSION TITLE
Portable Ruby Segfaults on OS X El Capitan - Cannot Install/Run Homebrew

### DISCUSSION THREAD

**Discussion by OldCrow** [Homebrew/discussions#5464] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.7
ORIGIN: https://github.com/Homebrew/brew
HEAD: 43eaeca50fe3b6c755f3fc9bc6f43669e0db1039
Last commit: 5 days ago
Core tap JSON: 01 Jan 05:00 UTC
Core cask tap JSON: 01 Jan 05:00 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 2
Homebrew Ruby: 3.3.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.3/bin/ruby
CPU: dual-core 64-bit penryn
Clang: 8.0.0 build 800
Git: 2.10.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.43.0 => /usr/bin/curl
macOS: 10.11.6-x86_64
CLT: 8.2.0.0.1.1480973914
Xcode: N/A

Note: /usr/bin/curl is used because Portable Ruby itself segfaults on all installs and Homebrew cannot install an updated curl.
```

### Output of `brew doctor`

```shell
Warning: Using `--insecure` with curl to download formula.jws.json because we need it to run `brew install curl` in order to download securely from now on. Checksums will still be verified.
Warning: Using `--insecure` with curl to download formula_tap_migrations.jws.json because we need it to run `brew install curl` in order to download securely from now on. Checksums will still be verified.
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You are using macOS 10.11.
We (and Apple) do not provide support for this old version.
It is expected behaviour that some formulae will fail to build in this old version.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.


Warning: An outdated version (2.10.1) of Git was detected in your PATH.
Git 2.14.3 or newer is required for Homebrew.
Please upgrade:
  brew install git

Note: As with curl, git cannot be updated as Portable Ruby segfaults.
```

### Description of issue

Running the install script on OS X 10.11 starts as expected with the creation of directories in /usr/local downloading and installing Homebrew from git and pouring the portable-ruby-3.3.3-el_capitan bottle. As this is El Capitan, curl needs to be downloaded because the system curl is outdated and insecure. The files for curl and it's dependencies download. On the first dependency build (pkg-config), Ruby segfaults. A file with the redirected output is attached.

Certainly there is no expectation the Homebrew supports any particular formula installation on 10.11 - the warning about 10.11 being outdated is clear. However, if Portable Ruby 3.3.3 has been built to provide an El Capitan bottle, then the Ruby installation itself should not segfault when running installation scripts. That's significantly different from a formula failing to build for whatever reason.
[brew.txt](https://github.com/user-attachments/files/16038383/brew.txt)

**Comment by Bo98** [comment]
Thanks for the logs - that's useful. Do you see any ruby crash report file in either `~/Library/Logs/DiagnosticReports` or `/Library/Logs/DiagnosticReports`? There might not be, but if there is please supply it!

Let's also try narrow this down. Can you run this and see if it crashes for you: `/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/current/bin/ruby -e 'p Random.urandom(4).unpack1("L")'`

Also try:
* `/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/current/bin/ruby -e 'Process.wait(fork { p Random.urandom(4).unpack1("L") })'`
* `/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/current/bin/ruby -e 'p Random.urandom(4).unpack1("L"); Process.wait(fork { p Random.urandom(4).unpack1("L") })'`

**Reply by Bo98** [reply]
cc @EricFromCanada - can you reproduce any of this with your 10.11 setup?

**Reply by OldCrow** [reply]
Sure thing - been playing around a bit so can't guarantee this is the crash that goes with the output I sent previously, but they all seem to be basically the same segfault on the child side of fork pre-exec.

Also attaching the output files from the two test commands.

[ruby_2024-06-28-210935-MacBook-Pro.crash.txt](https://github.com/user-attachments/files/16038572/ruby_2024-06-28-210935-MacBook-Pro.crash.txt)

[test1.txt](https://github.com/user-attachments/files/16038575/test1.txt)

[test2.txt](https://github.com/user-attachments/files/16038576/test2.txt)

**Reply by OldCrow** [reply]
Ah, sorry - I meant the three test commands and forgot to attach the first one, which is also the one that worked...
[test.txt](https://github.com/user-attachments/files/16038588/test.txt)

**Reply by EricFromCanada** [reply]
Same here. First command fine, next two: 💥 
[10.11_pkg-config_crash.txt](https://github.com/user-attachments/files/16039692/10.11_pkg-config_crash.txt)
[ruby_2024-06-29-062739_1011-El-Capitan.txt](https://github.com/user-attachments/files/16039693/ruby_2024-06-29-062739_1011-El-Capitan.txt)

**Reply by EricFromCanada** [reply]
As for 10.12, it seems to be working just fine with portable-ruby 3.3.3.

**Reply by Bo98** [reply]
Ruby 3.3 changed name generation for temporary files to use a cryptographic RNG (CCRandomGenerateBytes). CCRNG in 10.11 uses GCD under the hood which is incompatible in forks pre-exec. CCRNG was overhauled in 10.12 to avoid this issue.

It is perhaps somewhat telling it has taken over a month for anyone on 10.11 to notice this regression and there is potential for a future raise to 10.12 anyway in a recent discussion about `clonefile`. The days of 10.11 are perhaps numbered.

However it is possible to workaround this in the Ruby code by moving some things to be done ahead of forking. The code we have is not particularly safe to begin with and perhaps got lucky it even works on 10.12+. I'll investigate this later today.

**Reply by OldCrow** [reply]
Could very well be end of the road for 10.11 - in fact, I think I first saw this issue with Portable Ruby 3.3.1. But I had some other issues with my install of 10.11 and wiped/reinstalled the OS a few weeks ago. It was when this happened again on a near-pristine OS install that I started to get annoyed and decided to dig in a bit and then raise a support issue.

I'm not using Homebrew for much on this computer (the number of formulae that fail to build is always increasing as dependencies stop being built under the built-in Xcode tools and the built in is too old to build newer versions of llvm under Homebrew. But there are a few things I install and not being able to have them would be an inconvenience (but not an insurmountable one, of course. It would just make me a bit sad).

**Reply by OldCrow** [reply]
> Ruby 3.3 changed name generation for temporary files to use a cryptographic RNG (CCRandomGenerateBytes). CCRNG in 10.11 uses GCD under the hood which is incompatible in forks pre-exec. CCRNG was overhauled in 10.12 to avoid this issue.
> ...
> However it is possible to workaround this in the Ruby code by moving some things to be done ahead of forking. The code we have is not particularly safe to begin with and perhaps got lucky it even works on 10.12+. I'll investigate this later today.

Curious to know if this possible fix is likely to be implemented (or is implementable) - or if this is it for El Capitan?

**Reply by Bo98** [reply, marked as answer]
Apologies for the delay here - it wasn't a trivial fix and it was difficult to prioritise fixing it over other things given 10.11 usage is now at <0.01%.

I think I've fixed it in https://github.com/Homebrew/brew/pull/18183 which will be a part of 4.3.19.

@EricFromCanada Can you test to see if that PR works?

I don't think 10.11 support has too much of a future given the very low numbers but this will hopefully fix the issue until at least 4.4.0.

**Reply by EricFromCanada** [reply]
A fresh install on 10.11 of v4.3.18 currently fails at `==> Installing curl dependency: pkg-config`:
```
...
Error: Forked child process failed: pid 2828 SIGSEGV (signal 11)
Error: 'curl' must be installed and in your PATH!
Failed during: /usr/local/bin/brew update --force --quiet
1011-El-Capitan:~ vmadmin$ brew --config
HOMEBREW_VERSION: 4.3.18
ORIGIN: https://github.com/Homebrew/brew
HEAD: abc0584215659c5a8716d0798ea2801388d01368
Last commit: 2 days ago
Core tap JSON: 01 Jan 05:00 UTC
Core cask tap JSON: 01 Jan 05:00 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 3.3.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin/ruby
CPU: quad-core 64-bit ivybridge
Clang: 8.0.0 build 800
Git: 2.10.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.43.0 => /usr/bin/curl
macOS: 10.11.6-x86_64
CLT: 8.2.0.0.1.1480973914
Xcode: N/A
```

But if I update brew to the latest with `git`, then `brew update` seems to work.
```
1011-El-Capitan:~ vmadmin$ git -C `brew --repo` pull origin master -q
1011-El-Capitan:~ vmadmin$ brew update
...
==> Installing curl dependency: pkg-config
==> ./configure --disable-host-tool --with-internal-glib --with-pc-path=/usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/l
==> make
==> make install
...
```

**Reply by Bo98** [reply]
Yeah fresh installs should work once 4.3.19 is out. Thanks for checking!

**Reply by OldCrow** [reply]
Oh, wow - I figured this was just too far down the priority list to make the cut and/or too technically complicated to be worth the effort. This is really great! Looking forward to checking it out!
